src/BaseKrylov.fypp

```diff
@@ -121,84 +129,153 @@ contains
         endif
 
         return
-    end subroutine initialize_krylov_subspace_${t1[0]}$${k1}$
+    end subroutine initialize_krylov_subspace_${type[0]}$${kind}$
+
+    subroutine orthogonalize_vector_against_basis_${type[0]}$${kind}$(y, X, info, beta)
```

Comment:
A couple of questions about this utility function:

- The implementation seems to make the implicit assumption that `X` is an orthogonal basis to begin with. A test for orthogonality of `X` in the subroutine seems like overkill so, if my understanding is correct, we should make it very clear in the documentation of the function.
- `call linear_combination(proj, X, proj_coefficients) ; call y%sub(proj)` amounts to a plain iteration of the Gram-Schmit process. The implementation as it is is elegant and re-uses as many subroutines as possible but plain GS can be numerically unstable. Shouldn't we actually implement a modified GS iteration here to make sure we'll not run into any numerical accuracy problem eventually ?

Reply:
PS: I see that this function is being used in the `qr` where a two-pass strategy is used. I'm not quite sure the way the `orthogonalize_vector_against_...` is implemented is equivalent to the modified Gram Schmidt process where `y` in iteratively orthogonalized against columns of `X`. Here it seems like we have `batch` orthogonalization. Is this correct ?

Reply:
> * The implementation seems to make the implicit assumption that `X` is an orthogonal basis to begin with. A test for orthogonality of `X` in the subroutine seems like overkill so, if my understanding is correct, we should make it very clear in the documentation of the function.

Yes, this is correct, I also thought about that but opted against adding a orthonormality check on the basis for efficiency reasons, since this function is called many times. I will make it clear in the docs.

It's interesting in that what amounts to an orthogonal projection of `y` onto `X`, i.e. `XX^Ty` is the exact same code that I use in `LightROM` to compute `apply_outerprod` to compute the action of `BB^T`. Although it is tempting to reuse the code, I think it's confusing given that `B` is typically not orthonormal.

Originally, I had written the function to be `orthonormalize_against_basis` with an optional `ifnorm` flag, but I also rejected this option partly out of efficiency considerations because in the vast majority of calls we are not normalizing but request the projection coeffiicients as outputs.

> * `call linear_combination(proj, X, proj_coefficients) ; call y%sub(proj)` amounts to a plain iteration of the Gram-Schmit process. The implementation as it is is elegant and re-uses as many subroutines as possible but plain GS can be numerically unstable. Shouldn't we actually implement a modified GS iteration here to make sure we'll not run into any numerical accuracy problem eventually ?

Correct. In all the algorithms in which `orthogonalize_against_basis` is used for GS, it is indeed used twice. I considered extending it to double (or triple?) GS, but then I thought it might be good to keep it because there are moments where it's overkill. I would propose to add another routine called `double_gram-schmidt_step` which wraps two calls to `orthogonalize_against_basis` for clarity. 

It turns out this function is already written and is called `update_hessenberg` but is not exported. I will add this function in `BaseKrylov.f90` and remove `update_hessenberg`.

Reply:
> PS: I see that this function is being used in the `qr` where a two-pass strategy is used. I'm not quite sure the way the `orthogonalize_vector_against_...` is implemented is equivalent to the modified Gram Schmidt process where `y` in iteratively orthogonalized against columns of `X`. Here it seems like we have `batch` orthogonalization. Is this correct ?

No, it is not equivalent to the MGS. `orthogonalize_against_basis` is batch orthogonalization of a vector or a basis against an existing basis that is assumed orthonormal. The two are identical when only considering a single vector that needs to be orthogonalized against a basis. When considering a basis, on the other hand, then MGS requires us to orthogonalize each vector of the basis separately and add it to the basis, in such a way that the next vector is orthogonalized against both the original basis and the newest added vector, etc.

Reply:
> I would propose to add another routine called `double_gram-schmidt_step` which wraps two calls to `orthogonalize_against_basis` for clarity.

The only downside to this is that I and every user would need to actually remember (or repeatedly look up) how to spell Schmidt ...